### ISSUE TITLE
Allow setting editor-wide env vars from .gitpod.yml

### ISSUE THREAD

**Issue by TamaMcGlinn** [gitpod-io/gitpod#11928]
## Is your feature request related to a problem? Please describe
If you try to open any Alire based project in gitpod (for example, [Adabots_Examples](https://gitpod.io/#https://github.com/TamaMcGlinn/Adabots_examples)), the Ada language server extension will fail because it does not have the required environment variables set. Note how, although the Ada language server extension is installed in the above linked project, we get an error about missing .gpr files. This is because the environment variable GPR_PROJECT_PATH is not set up correctly. When opening a local editor, like VSCode, I can first open a terminal in the project root and issue `eval $(alr printenv)` to set all the environment variables as necessary, and then start vscode from that terminal. So I'm looking for a way to achieve the same using gitpod.

## Describe the behaviour you'd like
I want to be able to set a command to run in the workspace, before opening gitpod, such that they are defined for the whole workspace, including internal processes such as the LSP client inside a plugin.

## Describe alternatives you've considered
Alire itself has a nice workaround, but it's a hassle to have to create a special .gpr file that hardcodes the necessary relative paths just for gitpod. Alire itself needs this because otherwise you need alire to build alire, which would be annoying, but other alire projects should just assume that the user can set environment variables for the whole editor.

It would also be possible to fork the [Ada language server](https://github.com/AdaCore/ada_language_server) or the [Ada VSCode plugin](https://marketplace.visualstudio.com/items?itemName=AdaCore.ada), so that it would recognize Alire projects and set the environment variables itself. However, I'd rather not, since this isn't necessary for any other editors than gitpod, and I assume the respective project maintainers would therefore reject such a PR.

**Comment by TamaMcGlinn** [comment]
Note #9275 is similar, but I have tested and the approach given there only applies to the terminal windows, not to the editor as a whole and hence the language server still doesn't work.

**Comment by TamaMcGlinn** [comment]
I have also considered setting this inside a custom docker image, but the image would need the workspace mounted inside it in order to get this right. I don't think this is possible.

**Comment by axonasif** [comment]
Hi, `.gitpod.yml` does not support setting system-wide environment variables. It supports setting within the terminal scope. You can read more at https://www.gitpod.io/docs/environment-variables

What you're trying to achieve is best done from a dockerfile, it makes your workspace start faster too 🚀 

Raised a PR here: https://github.com/TamaMcGlinn/AdaBots_examples/pull/1

**Comment by jldec** [comment]
Things to try

1. add `eval $(alr printenv)` to the `before` task in gitpod.yml  
2. OR define the necessary env vars in your Gitpod project setting, or in your personal Gitpod settings.
3. OR set your env vars from a dotfile
4. OR (failing all 3 above) ask Gitpod team IDE

**Comment by TamaMcGlinn** [comment]
I have tried all of those. Number 4 is this issue.

**Comment by jldec** [comment]
I tried 1 in this PR and it appears to work
https://github.com/TamaMcGlinn/AdaBots_examples/pull/2

(the extension in gitpod.yml won't load, but the latest "AdaCore.ada" does load successfully with the change in the PR)

**Comment by TamaMcGlinn** [comment]
No, that doesn't work. $GPR_PROJECT_PATH gets set in the terminal, but not editor wide. Hence the error message that pops up about the project file having errors and being unable to load. Try opening any of the src/*.adb and see the red squiggly line below the first 'with *;' at the top of the file.

**Comment by jldec** [comment]
Assiged label `team: IDE`
I suspect that this is a problem with the extension, not with Gitpod.

**Comment by TamaMcGlinn** [comment]
Not sure what you mean; if you are saying there is something wrong with the Ada VSCode extension, I can confirm that this works in VSCode if you set the environment accordingly before launching VSCode. The error message in gitpod is exactly what you get locally in VSCode, when forgetting to set up the environment before starting VSCode.

![no_adabots](https://user-images.githubusercontent.com/5008897/185108920-3cc1b217-0312-4db8-a832-1c5ecf292de4.png)

This is because adabots.gpr is inside the gitignored alire/cache/dependencies directory, populated by running `alr build` or `alr printenv`, and finding that project file without a relative path to it is possible because of having GPR_PROJECT_PATH set accordingly.

As far as I understand, this problem is due to a fundamental gitpod design limitation. What I want to do is run 'alr printenv' from inside the working dir, but before gitpod has started the language server, so that the language server can see the environment variables and files set up in the working dir.

The only thing that should work given the current design of gitpod is to put your entire editor within the terminal, e.g. vim, since that editor is then started after setting up the environment variables. However, that gets pretty clunky, and the other terminal-related bugs in gitpod suddenly become major problems to getting anything done.

For example, you can fullscreen the gitpod terminal and use [AVE](https://github.com/TamaMcGlinn/ave), a dockerized Neovim setup for Ada:
```
docker run -it -e GPR_PROJECT_PATH=$GPR_PROJECT_PATH -v "$(pwd):/workspace/AdaBots_examples" --workdir=/workspace/AdaBots_examples tmcglinn/ave:latest
```

You can see the Ada language server works properly because we can jump to Adabots.Stack_Count (and because otherwise, the same 'imported project file not found' error message would have been present)

![peek_vim_in_gitpod](https://user-images.githubusercontent.com/5008897/185113974-61095a6f-b434-4b50-8f19-06e57fd555fc.gif)

**Comment by akosyakov** [comment]
It is indeed looks like a missing lifecycle to contribute env vars which depend on content but required for IDE. 

As a super hacky workaround, try to add following line in you dotfiles:
```
sed -i '/USER=gitpod/a eval $(alr printenv)' /ide/startup.sh
```

It changes startup code of VS Code, since .dotfiles are applied before it could work.

**Comment by TamaMcGlinn** [comment]
Wow @akosyakov after some modifications, that works!

You can try it yourself quite easily, by putting [[this url](https://github.com/TamaMcGlinn/gitpod_alire_dotfiles)](https://github.com/TamaMcGlinn/gitpod_alire_dotfiles) into your [preferences](https://gitpod.io/preferences) under Dotfiles repository url, and then [opening a new workspace of Adabots_Examples](https://gitpod.io/#https://github.com/TamaMcGlinn/Adabots_examples).

You will see that in any source file it is now possible to ctrl+click through to subprogram definitions. Great!

As a permanent solution, you could add a check in /ide/startup.sh for a special script in the workspace, say `.gitpod_pre_editor_startup.sh` and execute it if it exists.

**Comment by akosyakov** [comment]
> As a permanent solution, you could add a check in /ide/startup.sh for a special script in the workspace, say .gitpod_pre_editor_startup.sh and execute it if it exists. That would allow us to include it in the repo itself, and allows users to have dotfiles as well again.

We would need to think about it, i thought to add somewhere in .gitpod.yml i.e.`vscode.before` callback. I wonder also how it would work for VS Code Desktop. We don't control startup process there, i.e. VS Code Desktop connects and spawns SSH session and then propagate VS Code Sever over it. But it is another use case maybe.

**Comment by axonasif** [comment]
@akosyakov Doesn't the `supervisor` process inherit the environmental changes from a non-login interactive shell-config (i.e `.bashrc`) and hence the IDE processes as well being the child process?
<img width="1006" alt="Screenshot 2022-08-18 at 3 36 01 PM" src="https://user-images.githubusercontent.com/39482679/185363189-168dfebf-f104-44ee-b28c-c5aaaa2f4dd8.png">

If I export something from `.bashrc`, the IDE process would see it, right? (this was my observation, I had to export `PYTHONUSERBASE` from `.bashrc` with the context of my repo inside `/workspace` dir  to make [dbt-extension](https://github.com/innoverio/vscode-dbt-power-user) find my user-installed python packages)

In below picture, right side is the contents of `$HOME/.bashrc.d/60-python` which is exporting some `env` vars after the workspace starts with my dotfiles+repo-cloned and on the left I'm checking for their existence on each of the `/ide` process. So far, all if the `/ide` process got the variable.

<img width="1388" alt="Screenshot 2022-08-18 at 3 41 45 PM" src="https://user-images.githubusercontent.com/39482679/185364355-3e137ff5-3545-4ac1-9039-de01c7ffdd25.png">

**Comment by axonasif** [comment]
I usually solve these issues by writing a file inside `$HOME/.bashrc.d/` from a custom dockerfile, which would get loaded via our [`.bashrc` logic.](https://github.com/gitpod-io/workspace-images/blob/e91b47d148d6687703e258a7589b8cba74367a88/base/Dockerfile#L67)

**Comment by akosyakov** [comment]
> Doesn't the supervisor process inherit the environmental changes from a non-login interactive shell-config (i.e .bashrc) and hence the IDE processes as well being the child process?

It only inherits what defined in the workspace image.

> If I export something from .bashrc, the IDE process would see it, right? (this was my observation, I had to export PYTHONUSERBASE from .bashrc with the context of my repo inside /workspace dir to make [dbt-extension](https://github.com/innoverio/vscode-dbt-power-user) find my user-installed python packages)

it works for you, because PYTHONUSERBASE is not dependent on actual project content, but on content of the image. As in this case it is different.

**Comment by TamaMcGlinn** [comment]
In the PR, you add this to the dockerfile: `printf 'eval $(alr printenv)\n' > "$HOME/.bashrc.d/600-alr"`
but this actually needs to be something like `cd /workspace/AdaBots_examples; eval $(alr printenv)`

**Comment by axonasif** [comment]
> @TamaMcGlinn: In the PR, you add this to the dockerfile: printf 'eval $(alr printenv)\n' > "$HOME/.bashrc.d/600-alr"
but what we would need in the bashrc.d file is more like: cd /workspace/AdaBots_examples; eval $(alr printenv)

Ah, I see 👍. Didn't get the chance to test it myself, should've haha 😅 
Now I understand what you meant by "context of the repo", in that case you could swap that line with:
```dockerfile
RUN printf 'cd "$GITPOD_REPO_ROOT" && eval $(alr printenv)\n'
```

`GITPOD_REPO_ROOT` is a gitpod provided variable which points to your repo dir. (see this [page](https://www.gitpod.io/docs/environment-variables#default-environment-variables) for more info)

> @akosyakov: it works for you, because PYTHONUSERBASE is not dependent on actual project content, but on content of the image. As in this case it is different.

Got it, I found what I was missing earlier, thanks 👍

**Comment by akosyakov** [comment]
@axonasif fyi https://github.com/gitpod-io/gitpod/issues/10797

**Comment by TamaMcGlinn** [comment]
@jldec I've changed the title to reflect that this is a feature request, not a question about how to do things currently.

The workaround, modifying `/ide/startup.sh` inside your gitpod dotfiles, works. But it is very annoying to have to tell users they have to modify their dotfiles in order to use a particular repository. The whole idea of gitpod is that users don't need to configure anything to edit a given repository, so I am hoping for a better way to set editor-wide env vars from inside .gitpod.yml.

**Comment by akosyakov** [comment]
@TamaMcGlinn unfortunately it is a single request in very long time, and team is stretched on other things. PRs are welcomed if you are up to add such option.

Schema is here: https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-protocol/data/gitpod-schema.json
Startup code is here: https://github.com/gitpod-io/gitpod/blob/a2b7412db4a7d639a4458a83223b45d2aed1b52e/components/ide/code/startup.sh#L39
I probably will do it in code-helper itself, since you need to parse yml file and helper already does it. [[1](https://github.com/gitpod-io/gitpod/blob/a2b7412db4a7d639a4458a83223b45d2aed1b52e/components/ide/code/codehelper/main.go#L153)]

I would add it first as experimental similar to [[1](https://github.com/gitpod-io/gitpod/blob/a2b7412db4a7d639a4458a83223b45d2aed1b52e/components/gitpod-protocol/data/gitpod-schema.json#L253)] that we don't guarantee it support yet, but can enable you.

**Comment by jeanp413** [comment]
There's an issue in the ada language server about this https://github.com/AdaCore/ada_language_server/issues/668, @TamaMcGlinn can you comment in there :pray: 
we removed the startup.sh shell script so the workaorund mentioned above no longer works, closing this for now

**Comment by TamaMcGlinn** [comment]
That doesn't make any sense. ada_language_server is just one example usage of setting editor-wide environment variables, it's really not relevant to open an issue there about this, as ada_language_server works just fine after setting that variable. Since you're closing this, I assume you have found a way to set editor-wide env vars in gitpod?

**Comment by jeanp413** [comment]
Do you have another example about an use case for this? env vars are usually set in the dockerfile.
Regarding the case with Alire and Ada language, that should be fixed in ada lanaguage server as already reported in the linked issue

**Comment by TamaMcGlinn** [comment]
I don't normally use VSCode / gitpod, so I can't think of an example offhand, but it's a general feature request to be able to set env variables; you could go through other language servers to find any that might be affected by an environment variable, or find any VSCode plugin that might read environment variables; I don't think it would be very uncommon, but rather just accepted that it wouldn't work properly in gitpod.

Like I said, ada_language_server doesn't need fixing, it works fine and just requires certain env variables set if you want to configure the Ada environment; that works in local editors (NeoVim, VSCode, Emacs, GPS all handle this fine), it would be nice if it could work in gitpod as well.

It's fine that you don't want to fix it; but then you should close it as 'won't fix' rather than 'completed'.

Thanks for letting me know that the workaround doesn't work anymore. Is there an alternative to the startup.sh that would be similar?

**Comment by jeanp413** [comment]
>  you could go through other language servers to find any that might be affected by an environment variable, or find any VSCode plugin that might read environment variables; I don't think it would be very uncommon, but rather just accepted that it wouldn't work properly in gitpod.

Language servers for other languages work as expected, you can take a look at [this repo](https://github.com/gitpod-io/workspace-images/tree/main/chunks) and check how they are configured in the corresponding Dockerfile, you can also take a look at the ada language server repo as [they have a `gitpod.Dockerfile` configured](https://github.com/AdaCore/ada_language_server/blob/95a945d305c47d6799bbc85852d98da0d76183c4/.gitpod.Dockerfile#L2-L8)

>  ada_language_server doesn't need fixing, it works fine and just requires certain env variables set if you want to configure the Ada environment; that works in local editors (NeoVim, VSCode, Emacs, GPS all handle this fine)

It does need fixing, running `eval $(alr printenv)` as a prerequisite before starting vscode desktop so the ada language server can find the dependencies is not a good user experience, it should detect those dependencies automatically by scanning `alire.toml` as mentioned in the issue comments

>  Is there an alternative to the startup.sh that would be similar?

Not for now, as I mentioned, env variables can be set inside the Dockerfile or additionally you can set them in your [gitpod account settings](https://www.gitpod.io/docs/configure/projects/environment-variables#using-the-account-settings).
